File: moviejava/app/src/main/java/com/example/moviejava/data/remote/movie/MovieRepository.java
package com.example.moviejava.data.remote.movie;


import com.example.moviejava.model.JsonObject;
import com.example.moviejava.model.up_coming.UpComing;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.rxjava3.core.Observable;
import retrofit2.Retrofit;

@Singleton
public class MovieRepository {
    @Inject
    public MovieRepository(Retrofit retrofit) {
        this.service = retrofit.create(MovieService.class);
    }


    private MovieService service;

    public Observable<JsonObject> getTopRate(int page) {
        return service.getTopRate(page);
    }

    public Observable<UpComing> getUpcoming(int page) {
        return service.getUpcoming(page);
    }

    public Observable<JsonObject> getPopular(int page) {
        return service.getPopular(page);
    }

    public Observable<UpComing> getNowPlaying(int page) {
        return service.getNowPlaying(page);
    }

    public Observable<JsonObject> searchMovie(String query, int page) {
        return service.searchMovie(query, page);
    }
}

File: movie/app/src/main/java/com/example/movie/data/remote/movie/MovieRepository.kt
package com.example.movie.data.remote.movie

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import retrofit2.Retrofit
import javax.inject.Inject

class MovieRepository @Inject constructor(retrofit: Retrofit) {
    private val service = retrofit.create(MovieService::class.java)
    fun getUpcoming(page: Int) = flow {
        emit(
            service.getUpcoming(page = page)
        )
    }.flowOn(Dispatchers.IO)

    fun getTopRate(page: Int) = flow {
        emit(
            service.getTopRate(page = page)
        )
    }.flowOn(Dispatchers.IO)

    fun getNowPlaying(page: Int) = flow {
        emit(service.getNowPlaying(page = page))
    }.flowOn(Dispatchers.IO)

    fun getPopular(page: Int) = flow {
        emit(service.getPopular(page = page))
    }.flowOn(Dispatchers.IO)

    fun searchMovie(query: String, page: Int) = flow {
        emit(service.searchMovie(query = query, page = page))
    }.flowOn(Dispatchers.IO)
}